sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala

```diff
@@ -0,0 +1,107 @@
+package scalajsbundler.sbtplugin
+
+import org.scalajs.core.tools.io.FileVirtualJSFile
+import org.scalajs.sbtplugin.ScalaJSPlugin
+import org.scalajs.sbtplugin.ScalaJSPlugin.autoImport._
+import sbt.Keys._
+import sbt._
+
+import scalajsbundler._
+
+object NpmDepsPlugin extends AutoPlugin {
+
+  override lazy val requires = ScalaJSPlugin
+
+  // Exported keys
+  object autoImport {
+
+    /**
+      * Fetches NPM dependencies. Returns the directory in which the `npm install` command has been run.
+      *
+      * The plugin uses different directories according to the configuration (`Compile` or `Test`). Thus,
+      * this setting is scoped by a `Configuration`.
+      *
+      * Typically, if you want to define a task that uses the downloaded NPM packages you should
+      * specify the `Configuration` you are interested in:
+      *
+      * {{{
+      *   myCustomTask := {
+      *     val npmDirectory = (npmUpdate in Compile).value
+      *     doSomething(npmDirectory / "node_modules" / "some-package")
+      *   }
+      * }}}
+      *
+      * The task returns the directory in which the dependencies have been fetched (the directory
+      * that contains the `node_modules` directory).
+      *
+      * @group tasks
+      */
+    val npmUpdate: TaskKey[File] =
+      taskKey[File]("Fetch NPM dependencies")
+
+    /**
+      * Whether to use [[https://yarnpkg.com/ Yarn]] to fetch dependencies instead
+      * of `npm`. Yarn has a caching mechanism that makes the process faster.
+      *
+      * If set to `true`, it requires Yarn 0.22.0+ to be available on the
+      * host platform.
+      *
+      * Defaults to `false`.
+      *
+      * @group settings
+      */
+    val useYarn: SettingKey[Boolean] =
+      settingKey[Boolean]("Whether to use yarn for updates")
+  }
+
+  import autoImport._
+
+  val scalaJSBundlerPackageJson =
+    TaskKey[BundlerFile.PackageJson]("scalaJSBundlerPackageJson",
+      "Write a package.json file defining the NPM dependencies of project",
+      KeyRanks.Invisible
+    )
+
+  override lazy val projectSettings =
+    perScalaJSStageSettings(fullOptJS) ++ perScalaJSStageSettings(fastOptJS)
+
+  protected def perScalaJSStageSettings(stage: TaskKey[Attributed[File]]): Seq[Def.Setting[_]] = Seq(
+    useYarn := false
```

Comment:
In `ScalaJSBundlerPlugin`, `useYarn` used to be set in the project scope, directly in `projectSettings`. I think that's also where it should be in `NpmDepsPlugin`.

Speaking of which, the `useYarn := false` should be removed from `ScalaJSBundlerPlugin.projectSettings`, since it's redundant with the assignment in `NpmDepsPlugin.projectSettings`.

Reply:
In fact, you should inline the contents of your current `persScalaJSStageSettings` inside `projectSettings`, since you're not using `stage` at all here (which makes sense). So currently, you're just duplicating all these settings by calling `perScalaJSStageSettings(fullOptJS) ++ perScalaJSStageSettings(fastOptJS)`, for no reason.